### PR TITLE
Fix T587726: dxNumberBox - It's impossible to enter '0' in the decimal part after any digit is specified

### PR DESCRIPTION
**Pull request by mrjohn0011**

### CODE REVIEW

File: js/ui/number_box/number_box.mask.js [hunk 1]
```diff
@@ -275,25 +270,51 @@ var NumberBoxMask = NumberBoxBase.inherit({
...
-        return this._isPercentFormat() ? (parsed && parsed / 100) : parsed;
-    },
+        var precision = this._getMaxPrecision(format, parsed),
+            pow = Math.pow(10, precision);
 
-    _cleanText: function(text) {
-        var decimalSeparator = number.getDecimalSeparator(),
-            clearRegExp = new RegExp("[^0-9" + decimalSeparator + "]", "g"),
-            cleanedText = text.replace(clearRegExp, "").replace(decimalSeparator, ".");
+        parsed = Math.round((parsed === null ? this._parsedValue : parsed) * pow) / pow;
```
Comment (DokaRus) [inline js/ui/number_box/number_box.mask.js]: Extract this ternary expression. Operations with `pow` looks like a part of "false" branch expression of the ternary operator

File: js/ui/number_box/number_box.mask.js [hunk 2]
```diff
@@ -542,6 +569,13 @@ var NumberBoxMask = NumberBoxBase.inherit({
         return this.callBase.apply(this, arguments);
     },
 
+    _removeStubInText: function(text) {
+        var decimalSeparator = number.getDecimalSeparator(),
+            regExp = new RegExp("[^0-9" + decimalSeparator + "]+", "g");
```
Comment (DokaRus) [inline js/ui/number_box/number_box.mask.js:575]: Should we do that if `text` is defined?

File: js/ui/number_box/number_box.mask.js [hunk 3]
```diff
@@ -275,25 +270,51 @@ var NumberBoxMask = NumberBoxBase.inherit({
...
+        }
+
+        var clearedText = this._removeStubInText(text),
+            decimalSeparator = number.getDecimalSeparator(),
+            decimalSeparatorIndex = clearedText.indexOf(decimalSeparator),
+            separatorIsFirst = decimalSeparatorIndex === 0,
+            lastChar = clearedText.charAt(clearedText.length - 1),
+            onlyOneSeparatorExists = decimalSeparatorIndex === clearedText.length - 1;
+
+        if(separatorIsFirst) {
```
Comment (DokaRus) [inline js/ui/number_box/number_box.mask.js:305]: Can we combine conditions?
```
if(separatorIsFirst || lastChar !== "0")

Comment (mrjohn0011) [reply js/ui/number_box/number_box.mask.js:305]: No, we can't. lastChar may be not equal to zero but equal to decimalSeparator. In this case the method should return true:

```
if(lastChar === decimalSeparator && onlyOneSeparatorExists) {
   return true;
}
```

File: js/ui/number_box/number_box.mask.js (hunk 2 above)
Comment (mrjohn0011) [reply js/ui/number_box/number_box.mask.js:575]: Text cannot be undefined because it is always equal to `input().val()`

Comment (DokaRus) [reply js/ui/number_box/number_box.mask.js:575]:
```
 return (text || this._input().val()).replace(regExp, "");
```
long expression confused me, I've missed the first closing bracket. Can you make it more readable? :)

**Review by DokaRus — DISMISSED** [review]

**Review by DokaRus — APPROVED** [review]

**Pull request merged**